src/plugins/workspace/public/utils.ts

```
@@ -539,14 +596,32 @@ export const fetchDataSourceConnections = async (
 ) => {
   try {
     let directQueryConnections: DataSourceConnection[] = [];
+    let remoteClusterConnections: DataSourceConnection[] = [];
     if (mode === AssociationDataSourceModalMode.DirectQueryConnections) {
       directQueryConnections = await fetchDataSourceConnectionsByDataSourceIds(
         // Only data source saved object type needs to fetch data source connections, data connection type object not.
         dataSources.filter((ds) => ds.type === DATA_SOURCE_SAVED_OBJECT_TYPE).map((ds) => ds.id),
         http
       );
+    } else {
+      const remoteClusterConnectionsPromises = dataSources.map((ds) => {
+        if (
+          ds.dataSourceEngineType === DataSourceEngineType.OpenSearch ||
+          ds.dataSourceEngineType === DataSourceEngineType.Elasticsearch
+        ) {
+          return getRemoteClusterConnections(ds.id, http!).catch(() => []);
+        }
+        return [];
+      });
+      const remoteClusterConnectionsResult = await Promise.all(remoteClusterConnectionsPromises);
```

this might have the same effect if the underlying call doesn't timeout out these promises. on line 612, we are swallowing all errors so if it fails it will fail silently. and no error will be thrown.

then with promise all it will on be fulfilled when all of the input's promises fulfill. if we have 10 data sources it a single one of those data sources calls fails we will be left with dangling promises. for node 20 the garbage collection is extremely aggressive. so it will see these dangling promises left open and constantly take the time to prioritize gc and other parts of the application will get slowed down.

we should really avoid this pattern

---

I have updated the code to address your concerns regarding unfulfilled promises.